x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDataFrameAnalyticsAction.java

```diff
@@ -278,4 +336,58 @@ void waitForTaskRemoved(Set<String> analyticsIds, StopDataFrameAnalyticsAction.R
         return persistentTasks.findTasks(MlTasks.DATA_FRAME_ANALYTICS_TASK_NAME,
             t -> analyticsIds.contains(MlTasks.dataFrameAnalyticsIdFromTaskId(t.getId())));
     }
+
+    // Visible for testing
+    static class AnalyticsByTaskState {
+
+        final List<String> started;
+        final List<String> stopping;
+        final List<String> failed;
+
+        private AnalyticsByTaskState(List<String> started, List<String> stopping, List<String> failed) {
+            this.started = Collections.unmodifiableList(started);
+            this.stopping = Collections.unmodifiableList(stopping);
+            this.failed = Collections.unmodifiableList(failed);
+        }
+
+        boolean isEmpty() {
+            return started.isEmpty() && stopping.isEmpty() && failed.isEmpty();
+        }
+
+        List<String> getNonStopped() {
+            List<String> nonStopped = new ArrayList<>();
+            nonStopped.addAll(started);
+            nonStopped.addAll(stopping);
+            nonStopped.addAll(failed);
+            return nonStopped;
+        }
+
+        static AnalyticsByTaskState build(Set<String> analyticsIds, PersistentTasksCustomMetadata tasks) {
+            List<String> started = new ArrayList<>();
+            List<String> stopping = new ArrayList<>();
+            List<String> failed = new ArrayList<>();
+
+            for (String analyticsId : analyticsIds) {
+                switch (MlTasks.getDataFrameAnalyticsState(analyticsId, tasks)) {
+                    case STARTING:
+                    case STARTED:
+                    case REINDEXING:
+                    case ANALYZING:
+                        started.add(analyticsId);
+                        break;
+                    case STOPPING:
+                        stopping.add(analyticsId);
+                        break;
+                    case STOPPED:
+                        break;
+                    case FAILED:
+                        failed.add(analyticsId);
+                        break;
+                    default:
+                        break;
```

Comment:
I know you've only moved this in this PR, but maybe it's still best to add an `assert` in the `default` case just in case someone adds a new state but doesn't add it here?

Reply:
Done